platforms/web/lib/useListeners/useListeners.ts

```diff
@@ -188,9 +191,25 @@ export function useListeners(
 
         // this is required to handle edge case image pasting in Safari, see
         // https://github.com/vector-im/element-web/issues/25327
-        const onBeforeInput = onPaste;
+        const onBeforeInput = (e: ClipboardEvent | InputEvent) => {
+            if (isInputEvent(e) && e.isComposing) return;
+            return onPaste(e);
+        };
         editorNode.addEventListener('beforeinput', onBeforeInput);
 
+        const onCompositionEnd = (e: CompositionEvent): void => {
+            console.log(e);
+            // create a new inputEvent for us to process
+            const inputEvent = new InputEvent('input', {
+                data: e.data,
+                inputType: 'insertCompositionText',
+            });
+
+            // now process that new event
+            onInput(inputEvent);
```

Comment:
I had a look at the [`compositionend` event](https://developer.mozilla.org/en-US/docs/Web/API/Element/compositionend_event) and it sounds like this is also fired when the composition is cancelled.

If that happens, will we need to do something different?

Reply:
This is a good question, I think this is the correct behaviour.

For instance, on a mac, if you press `option + i` to insert a "composition style" ^ (ie the next letter will sit beneath the hat), then click off the textbox we _still_ process it as an input event, so that the `^` is being shown in the textbox.

To me that seems to be the correct behaviour for cancellation, but I'm not super familiar with this input method.